### PR TITLE
Improve error for unrecognized instructions

### PR DESCRIPTION
**Pull request by randall-fulton**

### CODE REVIEW

**Review by stylewarning — CHANGES_REQUESTED** [review]
nit: I think the condition should be exported

There are probably other places where this condition could be signaled. Have you done a quick scan through to see where it might be applicable? (I'm fine not being totally comprehensive in the first PR.)

Otherwise, this is a great idea and good addition. Thanks!

**Comment by randall-fulton** [comment]
@stylewarning Sorry for the delay. Had a nice little Covid battle there in the middle.

The condition is exported now. I've attempted to find any other obvious usages of the new condition, but am admittedly pretty unfamiliar with `quilc`, so it's more than possible I'm missing something. If you had anything specific in mind, let me know and I can make the change to introduce this condition.

File: src/addresser/initial-rewiring.lisp [hunk 1]
```diff
@@ -75,12 +85,14 @@ impossible using that rewiring.")
...
                   (application
                    ;; Merge the components of the arguments.
-                   (destructuring-bind (first . rest)
-                       (application-arguments inst)
-                     (ensure-qubit-component first)
-                     (dolist (qubit rest)
-                       (ensure-qubit-component qubit)
-                       (merge-qubit-components first qubit)))))))
+                   (let ((inst-arguments (application-arguments inst)))
+                     (when (zerop (list-length inst-arguments))
```
Comment (stylewarning) [inline src/addresser/initial-rewiring.lisp:89]: zerop list-length => endp

It's small but it changes it from linear to constant time.

**Review by stylewarning — CHANGES_REQUESTED** [review]
one tiny thing, then we are good to go!

File: src/addresser/initial-rewiring.lisp (hunk 1 above)
Comment (randall-fulton) [reply src/addresser/initial-rewiring.lisp:89]: Wonderful little nugget of knowledge. I'll remember that one.

**Review by notmgsk — CHANGES_REQUESTED** [review]

File: tests/addresser-tests.lisp [hunk 2]
```diff
@@ -27,6 +27,12 @@
                (cpp (quil::compiler-hook pp chip :protoquil t)))
           (is (quil::operator= orig-mat (quil:parsed-program-to-logical-matrix cpp))))))))
 
+(deftest test-unknown-instruction ()
+  (let ((program (safely-parse-quil "a"))
```
[the remark this answers is not in the document]
Comment (randall-fulton) [reply tests/addresser-tests.lisp:31]: 👀 Oh no

**Review by stylewarning — APPROVED** [review]

**Review by notmgsk — APPROVED** [review]

**Comment by notmgsk** [comment]
Thanks, @randall-fulton :D

**Comment by randall-fulton** [comment]
Thank you both for the feedback.

**Pull request merged**